### ISSUE TITLE
TRN-2 — Composable losses from observable specs, per-stage schedules, transform registry, and the SCF model-transform hook

### ISSUE THREAD

**Issue by aacostadiaz** [ACEsuit/mace#1576]
**Depends on:** TRN-1 (#1575), CORE-1 (#1555), ARCH-5 (#1633) · **Blocks:** GATE-3 (#1585), MAG-1 (#1584), TRN-5 (#1639), FT-4 (#1637), ELEC-3 (#1593)

**Prerequisite:** the Phase 0 characterization suite is committed — goldens, fixtures and the single-source tolerance table under `tests/golden/`, characterization tests under `tests/unit/`. References of the form `P0-N` name that work.

**Context:** Legacy losses are ten hardcoded `torch.nn.Module` classes in `mace/modules/loss.py` (659 lines): `WeightedEnergyForcesLoss` (`:246`), `WeightedForcesLoss` (`:272`), `WeightedEnergyForcesStressLoss` (`:290`), `WeightedHuberEnergyForcesStressLoss` (`:325`), `UniversalLoss` (`:391`), `WeightedEnergyForcesVirialsLoss` (`:506`), `DipoleSingleLoss` (`:543`), `DipolePolarLoss` (`:563`), `WeightedEnergyForcesDipoleLoss` (`:601`), `WeightedEnergyForcesL1L2Loss` (`:636`) — selected by string via `get_loss_fn(args, ...)` (`mace/tools/scripts_utils.py:772`, called at `mace/cli/run_train.py:834`), with the stage-two (ex-SWA) loss swapped mid-run inside the loop (`mace/tools/train.py:229-243`, `loss_fn = swa.loss_fn`). Missing-property masking already exists but as a helper the loss functions call by hand (`filter_nonzero_weight`, `mace/tools/utils.py:175`); adding a property loss still means forking a class. There is no transform layer (relative energies, masking schemes, inter–intra decomposition are all ad hoc or impossible); training schedules are limited to the single two-stage SWA switch plus the separate LBFGS optimizer swap; and SCF-style models have no sanctioned place to put a fixed-point solve with implicit differentiation — `MagneticSCFMACE` (`mace/modules/extensions.py:1976`) is a hand-written wrapper class that is not even reachable from `--model`. This ticket builds the v1 loss/transform/schedule layer that `TrainStage` (TRN-1 (#1575)) consumes: losses are **generated** from the declared observable specs (CORE-1 (#1555)), custom losses and data transforms are plugins in registries, optimizer/LR/loss are schedulable **per stage**, and models get a **model-transform hook** for SCF/implicit differentiation — a v1.0 must-have. The numeric behaviour to reproduce is pinned by P0-6's hand-computed loss characterization tests (`tests/unit/test_loss_characterization.py`), not by the legacy class structure.

**Interface & constraints:**

1. **Losses are generated from observable specs.** Every observable declared in config (CORE-1 (#1555) `ObservableSpec`: `name`, `irreps`, `per_atom`, `units`, `normalization`, `default_loss_weight`) automatically yields a loss term — a well-defined spherical-tensor observable needs **zero new loss code**. Derivative observables use CORE-1 (#1555)'s auto-naming (`d_<q>_d_pos` / `d_<q>_d_cell`, with `energy` special-cased to `forces` (sign −1) and `stress`) and get terms the same way. The total loss is the weighted linear combination of the per-observable terms. **A loss's parameters are config, not code:** the per-observable weights and a loss's own hyperparameters (e.g. `huber_delta` — legacy `WeightedHuberEnergyForcesStressLoss`/`UniversalLoss` at `mace/modules/loss.py:327,391` take it via `__init__`, wired from `--huber_delta` in `mace/tools/scripts_utils.py:800,808`) are read from `LossConfig` (weights + `params`); tuning an existing loss never means writing a new one.

2. **The numeric spec is P0-6, verbatim.** The generated losses must reproduce the hand-computed values pinned in `tests/unit/test_loss_characterization.py` for every legacy loss class's behaviour, including: per-atom vs per-config weighting semantics, `config_weight`, per-property weights, **missing-property masking** (a config lacking a labelled property contributes zero to that term, not NaN — legacy's `filter_nonzero_weight`, `mace/tools/utils.py:175`, and the per-property `property_weights` carried on each `Configuration`), and the `loss_weights` linear combination. Any deliberate semantic change must fail one of those tests and be argued in review — no silent renormalization.

3. **DDP-aware reduction is part of the loss contract, not an afterthought.** Legacy reduces raw losses through `reduce_loss` / `is_ddp_enabled` (`mace/modules/loss.py:19-48`), so a loss value under `--distributed` is not the single-process value of the same batch. Port that semantics explicitly: the generated loss states, in one place, how a term is reduced across ranks, and the CPU-DDP smoke (TRN-4 (#1578)) asserts a 2-rank loss matches the single-rank loss of the concatenated batch at fp64.

4. **Plugin loss registry.** Custom losses (non-standard reductions/transforms only) are `torch.nn.Module`s registered by decorator in `mace_torch/train/loss.py` — the decorator IS the registration, external packages register the same way, no registry edit:

```python
from mace_torch.train import register_loss

@register_loss("my_huber_ef")
class MyHuberEnergyForces(torch.nn.Module):
    def __init__(self, delta: float = 0.01):
        super().__init__()
        self.delta = delta

    def forward(self, pred, ref, weights):   # typed batch of MACEOutputs; returns a scalar
        e = torch.nn.functional.huber_loss(pred.total_energy, ref.total_energy, delta=self.delta)
        f = torch.nn.functional.huber_loss(pred.forces, ref.forces, delta=self.delta)
        return weights.energy * e + weights.forces * f
```

Selection is `LossConfig(name="my_huber_ef", weights=...)` (the `loss` config section, CFG-1 (#1574)). The signature is fixed: `forward(pred, ref, weights) -> scalar` over typed `MACEOutputs` batches — never raw dicts.

5. **Plugin transform registry + presets.** Data/loss transforms are registered via `@register_transform("name")` in `mace_torch/data/` and chained through `DataConfig(transforms=[...])`. Ship presets: **relative energies**, **masks**, **inter–intra decomposition**. Both per-atom and per-config quantities are supported. A **documented custom-transform example** (same altitude as the loss example above) is part of the deliverable — other schemes (absorption energies, many-body expansion, SO(3) magmom augmentation for MAG-1 (#1584)) must be expressible as registered transforms with zero core edits.

6. **Per-stage optimizer/LR/loss schedules.** Training is a sequence of stages; each stage may specify its optimizer, learning rate/scheduler, and loss (from the registry), with global settings as defaults. Two legacy special cases must be exactly expressible as stages, and both stop being loop-internal: the stage-two switch (`mace/tools/train.py:229-243`: swap loss, `AveragedModel.update_parameters`, `SWALR` step from `swa.start`) and the L-BFGS optimizer swap (`mace/cli/run_train.py:971-976`, executed by TRN-5 (#1639)). TRN-1 (#1575)'s explicit train/validate loop consumes the schedule; the schedule lives in the config's `schedule` section (CFG-1 (#1574)) and lands in the resolved-config metadata like everything else.

7. **Model-transform hook (SCF / implicit differentiation).** A model that solves a fixed point (charge equilibration, magnetic SCF) owns its fixed-point loop via a model-transform hook: the model is a `BaseMACE` subclass + `MACEOutputs` + the hook; inside the loop it calls dispatched ops per iteration (the long-range solver op is dispatched separately and is not this ticket's concern). The hook's contract: **force/stress training requires an implicit-differentiation backward** through the fixed point — an inference-only, non-differentiable solve is hard-rejected at build time for force training (loud error, never silently wrong gradients). Develop already learned this lesson the expensive way on the legacy wrapper: `MagneticSCFMACE.forward` accepts no `compute_magforces`, and `mace/cli/eval_configs.py:197-202` now refuses `--return_magforces` for an SCF-wrapped model up front instead of dying on a `TypeError` inside the wrapper — v1's contract must make that refusal structural. Downstream consumers: `MagneticSCFMACE` (MAG-1 (#1584)) and the `mace-scf` families `MACE-QEq` / `FixedPointSCF` (ELEC-3 (#1593)); this ticket ships the hook plus a toy fixed-point model proving the machinery, not those models.

**Task:**
1. Loss generation: observable spec → loss term factory; weighted combination; per-atom/per-config weighting, `config_weight`, property weights, missing-property masking matching P0-6 semantics exactly.
2. Cover every legacy loss class's behaviour through the generated path (or, where a legacy class is a genuinely non-standard reduction, as a registered loss) and check against the P0-6 hand-computed values.
3. Port the DDP reduction semantics of `mace/modules/loss.py:19-48` into the generated-loss contract and state it in one place.
4. `LOSS_REGISTRY` + `@register_loss` in `mace_torch/train/loss.py`; `LossConfig` selection wired into TRN-1 (#1575)'s `TrainStage`; the huber example above as documentation.
5. Transform registry (`@register_transform`, `DataConfig(transforms=[...])`) + the three presets (relative energies, masks, inter–intra) + the documented custom-transform example.
6. Per-stage schedules (optimizer/LR/loss per stage, global defaults) consumed by TRN-1 (#1575)'s loop; express legacy stage-two as a schedule and verify against the P0-5-pinned behaviour; expose the LBFGS stage kind TRN-5 (#1639) executes.
7. The model-transform hook: contract, build-time double-backward rejection, and a toy fixed-point model whose implicit-differentiation gradients are validated against unrolled autograd / finite differences.

**Out of scope:** multi-dataloader balancing and metrics/logging (TRN-3 (#1577)); checkpointing/DDP primitives (TRN-4 (#1578)); the L-BFGS closure and its DDP protocol (TRN-5 (#1639)); head creation and observables→heads (ARCH-3 (#1562)); the full training config schema (CFG-1 (#1574)); pseudolabel generation (FT-2 (#1581)); the actual SCF/magnetic/electrostatic models (ELEC-3 (#1593), MAG-1 (#1584)) and the long-range solver dispatch beneath them.

**Acceptance criteria:**
- [ ] Generated losses reproduce the P0-6 hand-computed values for all ten legacy loss class behaviours, including per-atom vs per-config weighting, `config_weight`, property weights, missing-property masking, and the `loss_weights` combination.
- [ ] A new spherical-tensor observable declared purely in config trains with an automatically generated loss term — zero new loss code (test demonstrates it end to end on a tiny fixture).
- [ ] A custom loss registered via `@register_loss` from a module outside `mace_torch` is selectable through `LossConfig` with no registry edit.
- [ ] The DDP reduction rule is stated once and asserted: a 2-rank gloo loss equals the single-rank loss of the concatenated batch at fp64.
- [ ] The three transform presets pass value tests; the documented custom-transform example runs as written.
- [ ] A two-stage schedule switches optimizer/LR/loss at the stage boundary; legacy stage-two (ex-SWA) semantics are reproduced as a schedule, not as loop-internal special casing.
- [ ] The toy fixed-point model's implicit-differentiation gradients agree with unrolled autograd/finite differences; a hook without a differentiable backward is hard-rejected at build time for force training.

**Inventory gaps assigned here:**
- [ ] Per-parameter-group LR factors as typed fields (replacing `--lr_params_factors`' JSON-in-a-string, parsed at `mace/tools/scripts_utils.py:942`), including the `--freeze` path that reuses them by zeroing factors.
- [ ] `--magforces_weight` as an observable weight and `--stage_two_magforces_weight` as a per-stage schedule entry (the `--swa_magforces_weight` option-string spelling is a recorded DROP; both spellings share the `swa_magforces_weight` dest at `mace/tools/arg_parser.py:809-815`).
- [ ] `--config_type_weights` — the per-`config_type` weighting that feeds `Configuration.weight`; no test asserts a non-default value changes a loss.

**Verify:**
```bash
python -m pytest packages/mace-torch/tests/train/test_loss_generation.py packages/mace-torch/tests/train/test_loss_registry.py -v
python -m pytest packages/mace-torch/tests/data/test_transforms.py packages/mace-torch/tests/train/test_stage_schedules.py -v
python -m pytest packages/mace-torch/tests/train/test_model_transform_hook.py -v
python -m pytest tests/unit/test_loss_characterization.py -v   # P0-6 spec, frozen legacy values
```
**Review focus:** loss weighting/masking semantics against the P0-6 arithmetic (the silent-renormalization risk), the DDP reduction rule, and the implicit-differentiation backward of the model-transform hook — the last two need a physics reviewer.

---
